main/modes/mode_paint.c

```diff
@@ -0,0 +1,2266 @@
+    { .name = "Circle",     .mode = PICK_POINT, .maxPoints = 2, .minSize = 1, .maxSize = 1, .fnDraw = paintDrawCircle },
+    { .name = "Filled Circle", .mode = PICK_POINT, .maxPoints = 2, .minSize = 0, .maxSize = 0, .fnDraw = paintDrawFilledCircle },
+    { .name = "Ellipse",    .mode = PICK_POINT, .maxPoints = 2, .minSize = 1, .maxSize = 1, .fnDraw = paintDrawEllipse },
+    { .name = "Polygon",    .mode = PICK_POINT_LOOP, .maxPoints = MAX_PICK_POINTS, .minSize = 1, .maxSize = 1, .fnDraw = paintDrawPolygon },
+    { .name = "Squarewave", .mode = PICK_POINT, .maxPoints = 2, .minSize = 1, .maxSize = 1, .fnDraw = paintDrawSquareWave },
+    { .name = "Paint Bucket", .mode = PICK_POINT, .maxPoints = 1, .minSize = 0, .maxSize = 0, .fnDraw = paintDrawPaintBucket },
+    { .name = "Clear",      .mode = INSTANT, .maxPoints = 0, .minSize = 0, .maxSize = 0, .fnDraw = paintDrawClear },
+};
+
+paintMenu_t* paintState;
+
+// Util function declarations
+
+int xTranslate(int x);
+int yTranslate(int y);
+paletteColor_t getContrastingColor(paletteColor_t col);
+void paintInitialize();
+void paintRenderAll();
+void restoreCursorPixels();
+void paintRenderCursor();
+void paintRenderToolbar();
+void paintDoTool(uint16_t x, uint16_t y, paletteColor_t col);
+void paintMoveCursorRel(int8_t xDiff, int8_t yDiff);
+void paintClearCanvas();
+void paintUpdateRecents(uint8_t selectedIndex);
+void paintDrawPickPoints();
+void paintHidePickPoints();
+void paintDrawWsgTemp(display_t* display, wsg_t* wsg, pxStack_t* saveTo, uint16_t x, uint16_t y, colorMapFn_t colorSwap);
+void paintSetupTool();
+void paintPrevTool();
+void paintNextTool();
+void paintLoadIndex();
+void paintSaveIndex();
+void paintResetStorage();
+bool paintGetSlotInUse(uint8_t slot);
+void paintSetSlotInUse(uint8_t slot);
+uint8_t paintGetRecentSlot();
+void paintSetRecentSlot(uint8_t slot);
+void paintSave(uint8_t slot);
+void paintLoad(uint8_t slot);
+void floodFill(display_t* disp, uint16_t x, uint16_t y, paletteColor_t col);
+void _floodFill(display_t* disp, uint16_t x, uint16_t y, paletteColor_t search, paletteColor_t fill);
+void _floodFillInner(display_t* disp, uint16_t x, uint16_t y, paletteColor_t search, paletteColor_t fill);
+
+void initPxStack(pxStack_t* pxStack);
+void freePxStack(pxStack_t* pxStack);
+void maybeGrowPxStack(pxStack_t* pxStack);
+void maybeShrinkPxStack(pxStack_t* pxStack);
+void pushPx(pxStack_t* pxStack, display_t* disp, uint16_t x, uint16_t y);
+bool popPx(pxStack_t* pxStack, display_t* disp);
+
+
+// Generic util functions
+
+int xTranslate(int x)
+{
+    return (x - PAINT_CANVAS_X_OFFSET) * PAINT_CANVAS_SCALE + PAINT_CANVAS_X_OFFSET + paintState->subPixelOffsetX;
+}
+
+int yTranslate(int y)
+{
+    return (y - PAINT_CANVAS_Y_OFFSET) * PAINT_CANVAS_SCALE + PAINT_CANVAS_Y_OFFSET + paintState->subPixelOffsetY;
+}
+
+void paletteColorToLed(paletteColor_t pal, led_t* led)
```

Comment:
The math in `paletteColorToLed()` and `ledColorToPalette()` feels weird. 51?

Reply:
FWIW, this function already exists in `swadge_util.c` (though maybe not in your branch?
```C
paletteColor_t RGBtoPalette( uint32_t rgb )
{
    uint8_t r = (rgb >>  0) & 0xFF;
    uint8_t g = (rgb >>  8) & 0xFF;
    uint8_t b = (rgb >> 16) & 0xFF;
    r = (r * 6) / 256;
    g = (g * 6) / 256;
    b = (b * 6) / 256;
    return (paletteColor_t) ( r * 36) + (g * 6) + b;
}
```
In reverse
```C
b = ((palette % 6) * 256) / 6;
g = (((palette / 6) % 6) * 256) / 6;
r = ((palette / 36) * 256) / 6;
```

Reply:
Ah, so it does! I should merge in main more often.